plugins/out_websocket/websocket.c

```diff
@@ -0,0 +1,283 @@
+    
+    if (!u_conn) {
+        flb_error("[out_ws] no upstream connections available to %s:%i", u->tcp_host, u->tcp_port);
+        ctx->handShake = true;
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+    now = time(NULL);
+
+    //TODO how to determine the interval? conn disconnet is about 30 sec, so we set 20 ssecnds here. 
+    flb_debug("[out_ws] interval is  %ld and handShake is %s", now - ctx->lastInputTimestamp, ctx->handShake?"true":"false");
+    if((now - ctx->lastInputTimestamp > ctx->idle_interval) && (ctx->handShake == false)) {
+        ctx->handShake = true;
+        flb_upstream_conn_release(u_conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    } 
+    ctx->lastInputTimestamp = now;
+
+    if(ctx->handShake == true) {
+        /* Handshake with websocket server*/
+        flb_info("[out_ws] handshake for ws");
+        ret = flb_ws_handshake(u_conn, ctx);
+        if (ret == -1) {
+            flb_upstream_conn_release(u_conn);
+            FLB_OUTPUT_RETURN(FLB_RETRY);
+        }
+        ctx->handShake = false;
+    }
+
+    /* Data format process*/
+    if (ctx->out_format != FLB_PACK_JSON_FORMAT_NONE) {
+        json = flb_pack_msgpack_to_json_format(data, bytes,
+                                               ctx->out_format,
+                                               ctx->json_date_format,
+                                               ctx->json_date_key);
+        
+        if (!json) {
+            flb_error("[out_ws] error formatting JSON payload");
+            flb_upstream_conn_release(u_conn);
+            FLB_OUTPUT_RETURN(FLB_ERROR);
+        }
+    }
+
+    /* Write message header */
+    if (ctx->out_format == FLB_PACK_JSON_FORMAT_NONE) {
+        ret = flb_ws_sendDataFrameHeader(u_conn, ctx, data, bytes);
+    }
+    else {
+        ret = flb_ws_sendDataFrameHeader(u_conn, ctx, json, flb_sds_len(json));
+    }
+    if (ret == -1) {
+        flb_error("[out_ws] dataFrameHeader sent failed");
+        flb_upstream_conn_release(u_conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+    /* Write message body*/
+    if (ctx->out_format == FLB_PACK_JSON_FORMAT_NONE) {
+        ret = flb_io_net_write(u_conn, data, bytes, &bytes_sent);
+    }
+    else {
+        ret = flb_io_net_write(u_conn, json, flb_sds_len(json), &bytes_sent);
+        flb_sds_destroy(json);
+    }
+
+    //flb_info("[out_ws] sendDataFrame number of bytes sent = %i", ret);
+    if (ret == -1) {
+        flb_upstream_conn_release(u_conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+	
+    /* Release the connection */
+    flb_upstream_conn_release(u_conn);
+    FLB_OUTPUT_RETURN(FLB_OK);
+}
+
+/* Plugin reference */
+struct flb_output_plugin out_websocket_plugin = {
+    .name         = "websocket",
+    .description  = "Websocket",
+    .cb_init      = cb_ws_init,
+    .cb_flush     = cb_ws_flush,
+    .cb_exit      = cb_ws_exit,
+    .flags        = FLB_OUTPUT_NET | FLB_IO_OPT_TLS | FLB_IO_TCP_KA,
```

Comment:
the FLB_IO_TCP_KA flag is not necessary, by default all TCP connections aims to be keepalive enabled by default